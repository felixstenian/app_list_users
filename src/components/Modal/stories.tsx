import { Meta, Story } from '@storybook/react'
import Modal from '.'

export default {
  title: 'Modal',
  component: Modal,
  args: {}
} as Meta

export const Default: Story = (args) => <Modal {...args} />
