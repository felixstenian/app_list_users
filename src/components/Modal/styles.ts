import styled, { css } from 'styled-components'

import { Row } from 'components'
import { RowProps } from 'components/Row'

export const Wrapper = styled(Row)<{ isOpen: boolean }>`
  ${({ theme, isOpen }) => css`
    width: 100%;
    height: 100%;
    top: 0;
    bottom: 0;
    position: fixed;
    display: ${isOpen ? 'flex' : 'none'};
    justify-content: center;
    z-index: ${theme.layers.modal};
    background: ${theme.colors.overlay.light};
  `}
`

export const Container = styled(Row)<RowProps>`
  flex-direction: column;
  position: relative;
  min-width: 400px;
  margin-left: 24px;
  margin-right: 24px;
  padding: 24px;
  border-radius: 8px;
  background-color: white;
`

export const CloseIcon = styled(Row)`
  position: absolute;
  top: 24px;
  right: 24px;
  background: none;
  cursor: pointer;
`
