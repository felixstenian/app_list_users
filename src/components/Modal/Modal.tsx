import { AiOutlineCloseCircle } from 'react-icons/ai'
import { LayoutProps, SpaceProps } from 'styled-system'

import { RowProps } from 'components/Row'

import * as Style from './styles'

export interface ModalProps extends RowProps, LayoutProps, SpaceProps {
  hasCloseIcon?: boolean
  open: boolean
  onClose: () => void
}

const Modal = ({
  hasCloseIcon = false,
  open,
  children,
  onClose,
  ...rest
}: React.PropsWithChildren<ModalProps>) => {
  const closeFunction = () => {
    onClose()
  }

  return (
    <>
      {open && (
        <Style.Wrapper isOpen={open}>
          <Style.Container {...rest}>
            {hasCloseIcon && (
              <Style.CloseIcon
                onClick={closeFunction}
                data-testid="modal-close-icon"
              >
                <AiOutlineCloseCircle size={20} color="gray.600" />
              </Style.CloseIcon>
            )}
            {children}
          </Style.Container>
        </Style.Wrapper>
      )}
    </>
  )
}

export default Modal
