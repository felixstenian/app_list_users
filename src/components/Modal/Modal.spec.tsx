import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils'

import Modal from '.'

describe('<Modal />', () => {
  it('1. should render the reading', () => {
    const { container } = renderWithTheme(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      <Modal open={true} onClose={() => {}}>
        Modal
      </Modal>
    )

    expect(screen.getByText(/Moda/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('2. should render the modal with close icon when hasCloseIcon is true', () => {
    renderWithTheme(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      <Modal open={true} onClose={() => {}} hasCloseIcon={true} />
    )

    expect(screen.getByTestId('modal-close-icon')).toBeInTheDocument()
  })
  it('3. should call onClose function when close icon is clicked', () => {
    const mockOnClose = jest.fn()
    const { getByTestId } = renderWithTheme(
      <Modal open={true} onClose={mockOnClose} hasCloseIcon={true} />
    )
    fireEvent.click(getByTestId('modal-close-icon'))
    expect(mockOnClose).toHaveBeenCalled()
  })
})
