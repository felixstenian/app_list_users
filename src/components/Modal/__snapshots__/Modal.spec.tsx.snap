// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> 1. should render the reading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 100%;
  height: 100%;
  top: 0;
  bottom: 0;
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  z-index: 40;
  background: rgba(0,0,0,0.25);
}

.c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 400px;
  margin-left: 24px;
  margin-right: 24px;
  padding: 24px;
  border-radius: 8px;
  background-color: white;
}

<div
  class="c0 c1"
>
  <div
    class="c0 c2"
  >
    Modal
  </div>
</div>
`;
